lib/ecto/query.ex

```diff
@@ -1446,12 +1468,17 @@ defmodule Ecto.Query do
   '''
   defmacro with_cte(query, name, opts) do
     with_query = opts[:as]
+    operation = opts[:operation]
 
     if !with_query do
       Builder.error!("`as` option must be specified")
     end
+    
+    unless operation in [nil, :all, :update_all, :delete_all, :insert_all] do
+      Builder.error!("`operation` option must be one of :all, :update_all, :delete_all, :insert_all")
+    end
```

Comment:
This validation must happen inside `apply`. Here the validation is too early (if it is inside the macro, it is happening at compile-time). We should validate and convert `nil` to `:all` inside CTE.apply. :)

Reply:
I've moved it inside `apply` but now it's referencing the wrong place, hinting at the query instead of the `with_cte/3` call. Are we OK with that? It still feels to me like this should be a compile-time error rather than a runtime one, on invalid option?

```
** (Ecto.Query.CompileError) `operation` option must be one of :all, :update_all, :delete_all, :insert_all
     code: |> select([c], %{id: c.id, desc: c.desc})
     stacktrace:
       test/ecto/adapters/postgres_test.exs:398: (test)
```